source/common/unified_malloc_framework/src/memory_tracker.c

```diff
@@ -9,112 +9,117 @@
  */
 
 #include "memory_tracker.h"
+#include "critnib/critnib.h"
+
 #include <umf/memory_provider.h>
 #include <umf/memory_provider_ops.h>
 
-#include <cassert>
-#include <map>
-#include <mutex>
-#include <shared_mutex>
+#include <assert.h>
+#include <errno.h>
 #include <stdlib.h>
 
 #ifdef _WIN32
 #include <windows.h>
 #endif
 
-// TODO: reimplement in C and optimize...
-struct umf_memory_tracker_t {
-    enum umf_result_t add(void *pool, const void *ptr, size_t size) {
-        std::unique_lock<std::shared_mutex> lock(mtx);
+#if defined(_WIN32) && defined(UMF_SHARED_LIBRARY)
+critnib *TRACKER = NULL;
+BOOL APIENTRY DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved) {
+    if (fdwReason == DLL_PROCESS_DETACH) {
+        critnib_delete(TRACKER);
+    } else if (fdwReason == DLL_PROCESS_ATTACH) {
+        TRACKER = critnib_new();
+    }
+    return TRUE;
+}
+#elif defined(_WIN32)
+struct tracker_t {
+    tracker_t() { map = critnib_new(); }
+    ~tracker_t() { critnib_remove(map); }
+    critnib *map;
+};
+tracker_t TRACKER_INSTANCE;
+critnib *TRACKER = TRACKER_INSTANCE.map;
+#else
+critnib *TRACKER = NULL;
+void __attribute__((constructor)) createLibTracker() {
+    TRACKER = critnib_new();
+}
+void __attribute__((destructor)) deleteLibTracker() { critnib_delete(TRACKER); }
+#endif
 
-        if (size == 0) {
-            return UMF_RESULT_SUCCESS;
-        }
+umf_memory_tracker_handle_t umfMemoryTrackerGet(void) {
+    return (umf_memory_tracker_handle_t)TRACKER;
+}
 
-        auto ret =
-            map.try_emplace(reinterpret_cast<uintptr_t>(ptr), size, pool);
-        return ret.second ? UMF_RESULT_SUCCESS : UMF_RESULT_ERROR_UNKNOWN;
-    }
+struct tracker_value {
+    void *pool;
+    size_t size;
+};
 
-    enum umf_result_t remove(const void *ptr, size_t size) {
-        std::unique_lock<std::shared_mutex> lock(mtx);
+static enum umf_result_t
+umfMemoryTrackerAdd(umf_memory_tracker_handle_t hTracker, void *pool,
+                    const void *ptr, size_t size) {
+    assert(ptr);
 
-        map.erase(reinterpret_cast<uintptr_t>(ptr));
+    struct tracker_value *value =
+        (struct tracker_value *)malloc(sizeof(struct tracker_value));
+    value->pool = pool;
+    value->size = size;
 
-        // TODO: handle removing part of the range
-        (void)size;
+    int ret = critnib_insert((critnib *)hTracker, (uintptr_t)ptr, value, 0);
```

Comment:
Why can we not use `void*` as a key and should use `uintptr_t` instead?

Reply:
Sorry, I forgot that it is not a C++ where we can use the power of metaprogramming. :)